alf/environments/parallel_environment.py

```diff
@@ -155,14 +194,68 @@ def _step(self, actions):
         Returns:
             Batch of observations, rewards, and done flags.
         """
-        time_steps = [
-            env.step(action, self._blocking)
-            for env, action in zip(self._envs, self._unstack_actions(actions))
-        ]
+        if not self._blocking:
+            time_steps = self._step_or_handle_last_done(
+                self._unstack_actions(actions))
+        else:
+            time_steps = [
+                env.step(action, self._blocking) for env, action in zip(
+                    self._envs, self._unstack_actions(actions))
+            ]
+
         # When blocking is False we get promises that need to be called.
         if not self._blocking:
             time_steps = [promise() for promise in time_steps]
-        return self._stack_time_steps(time_steps)
+
+        time_steps = self._stack_time_steps(time_steps)
+
+        if not self._blocking:
+            self._handle_done(time_steps)
+
+        return time_steps
+
+    def _step_or_handle_last_done(self, actions):
+        time_steps = []
+        for i, (env, action) in enumerate(zip(self._envs, actions)):
+            reset_ts = self._reset_ts[i]
+            if reset_ts is not None:
+                # reset was already called, simply use the result from previous stepping
+                time_steps.append(reset_ts)
+                self._reset_ts[i] = None
+            else:
+                time_steps.append(env.step(action, self._blocking))
+        return time_steps
+
+    def _handle_done(self, time_steps):
+        for i, reset_ts in enumerate(self._reset_ts):
+            if reset_ts is not None:
```

Comment:
How can this happen?

Reply:
I was handling the _reset() lazily here.  Now it's no longer needed.